File: crates/ifc-lite-yew/src/components/properties_panel.rs
//! Properties panel - shows selected entity details

use std::collections::HashSet;
use yew::prelude::*;
use crate::state::{ViewerAction, ViewerStateContext};

/// Properties panel component
#[function_component]
pub fn PropertiesPanel() -> Html {
    let state = use_context::<ViewerStateContext>().expect("ViewerStateContext not found");

    // Get selected entity
    let selected_entity = state.selected_ids.iter().next()
        .and_then(|id| state.entities.iter().find(|e| e.id == *id));

    html! {
        <div class="properties-panel">
            if let Some(entity) = selected_entity {
                // Entity info section
                <div class="property-section">
                    <div class="section-header">{"Entity Info"}</div>

                    <div class="property-row">
                        <span class="property-label">{"Type"}</span>
                        <span class="property-value">{&entity.entity_type}</span>
                    </div>

                    if let Some(ref name) = entity.name {
                        <div class="property-row">
                            <span class="property-label">{"Name"}</span>
                            <span class="property-value">{name}</span>
                        </div>
                    }

                    if let Some(ref global_id) = entity.global_id {
                        <div class="property-row">
                            <span class="property-label">{"GlobalId"}</span>
                            <span class="property-value global-id">
                                {global_id}
                                <button
                                    class="copy-btn"
                                    onclick={
                                        let gid = global_id.clone();
                                        Callback::from(move |_| {
                                            copy_to_clipboard(&gid);
                                        })
                                    }
                                    title="Copy to clipboard"
                                >
                                    {"📋"}
                                </button>
                            </span>
                        </div>
                    }

                    if let Some(ref storey) = entity.storey {
                        <div class="property-row">
                            <span class="property-label">{"Storey"}</span>
                            <span class="property-value">{storey}</span>
                        </div>
                    }

                    if let Some(elevation) = entity.storey_elevation {
                        <div class="property-row">
                            <span class="property-label">{"Elevation"}</span>
                            <span class="property-value">{format!("{:.2} m", elevation)}</span>
                        </div>
                    }
                </div>

                // Actions section
                <div class="property-section">
                    <div class="section-header">{"Actions"}</div>

                    <div class="action-buttons">
                        <button
                            class="action-btn"
                            onclick={
                                let entity_id = entity.id;
                                Callback::from(move |_| {
                                    crate::bridge::save_focus(&crate::bridge::FocusData { entity_id });
                                    crate::bridge::log(&format!("Zoom to entity #{}", entity_id));
                                })
                            }
                            title="Zoom to entity"
                        >
                            {"🔍 Zoom to"}
                        </button>

                        <button
                            class="action-btn"
                            onclick={
                                let state = state.clone();
                                let entity_id = entity.id;
                                Callback::from(move |_| {
                                    state.dispatch(ViewerAction::IsolateEntity(entity_id));
                                })
                            }
                            title="Isolate entity"
                        >
                            {"🎯 Isolate"}
                        </button>

                        <button
                            class="action-btn"
                            onclick={
                                let state = state.clone();
                                let entity_id = entity.id;
                                Callback::from(move |_| {
                                    state.dispatch(ViewerAction::HideEntity(entity_id));
                                })
                            }
                            title="Hide entity"
                        >
                            {"👁‍🗨 Hide"}
                        </button>

                        <button
                            class="action-btn"
                            onclick={
                                let state = state.clone();
                                let entity_type = entity.entity_type.clone();
                                Callback::from(move |_| {
                                    // Select all entities of the same type
                                    let same_type_ids: HashSet<u64> = state.entities.iter()
                                        .filter(|e| e.entity_type == entity_type)
                                        .map(|e| e.id)
                                        .collect();
                                    for id in same_type_ids {
                                        state.dispatch(ViewerAction::AddToSelection(id));
                                    }
                                })
                            }
                            title="Select all of this type"
                        >
                            {"📑 Select Similar"}
                        </button>
                    </div>
                </div>

                // Property Sets
                if !entity.property_sets.is_empty() {
                    { for entity.property_sets.iter().map(|pset| html! {
                        <div class="property-section">
                            <div class="section-header">{&pset.name}</div>
                            { for pset.properties.iter().map(|prop| html! {
                                <div class="property-row">
                                    <span class="property-label">{&prop.name}</span>
                                    <span class="property-value">
                                        {&prop.value}
                                        if let Some(ref unit) = prop.unit {
                                            <span class="property-unit">{format!(" {}", unit)}</span>
                                        }
                                    </span>
                                </div>
                            })}
                        </div>
                    })}
                } else {
                    <div class="property-section">
                        <div class="section-header">{"Property Sets"}</div>
                        <div class="empty-state small">
                            <span class="empty-text">{"No property sets"}</span>
                        </div>
                    </div>
                }

                // Quantities
                if !entity.quantities.is_empty() {
                    <div class="property-section">
                        <div class="section-header">{"Quantities"}</div>
                        { for entity.quantities.iter().map(|qty| html! {
                            <div class="property-row">
                                <span class="property-label">{&qty.name}</span>
                                <span class="property-value">
                                    {format!("{:.3}", qty.value)}
                                    if !qty.unit.is_empty() {
                                        <span class="property-unit">{format!(" {}", qty.unit)}</span>
                                    }
                                </span>
                            </div>
                        })}
                    </div>
                } else {
                    <div class="property-section">
                        <div class="section-header">{"Quantities"}</div>
                        <div class="empty-state small">
                            <span class="empty-text">{"No quantities"}</span>
                        </div>
                    </div>
                }
            } else if state.selected_ids.len() > 1 {
                // Multiple selection
                <div class="multi-selection">
                    <span class="selection-icon">{"📑"}</span>
                    <span class="selection-count">
                        {format!("{} entities selected", state.selected_ids.len())}
                    </span>

                    <div class="action-buttons">
                        <button
                            class="action-btn"
                            onclick={
                                let state = state.clone();
                                Callback::from(move |_| {
                                    let ids = state.selected_ids.clone();
                                    state.dispatch(ViewerAction::IsolateEntities(ids));
                                })
                            }
                        >
                            {"🎯 Isolate All"}
                        </button>

                        <button
                            class="action-btn"
                            onclick={
                                let state = state.clone();
                                Callback::from(move |_| {
                                    for id in state.selected_ids.iter() {
                                        state.dispatch(ViewerAction::HideEntity(*id));
                                    }
                                })
                            }
                        >
                            {"👁‍🗨 Hide All"}
                        </button>

                        <button
                            class="action-btn"
                            onclick={
                                let state = state.clone();
                                Callback::from(move |_| {
                                    state.dispatch(ViewerAction::ClearSelection);
                                })
                            }
                        >
                            {"✖ Clear Selection"}
                        </button>
                    </div>
                </div>
            } else {
                // No selection
                <div class="empty-state">
                    <span class="empty-icon">{"👆"}</span>
                    <span class="empty-text">{"No entity selected"}</span>
                    <span class="empty-hint">{"Click an entity to view its properties"}</span>
                </div>
            }
        </div>
    }
}

/// Copy text to clipboard using JS eval
fn copy_to_clipboard(text: &str) {
    // Simple approach using JS eval
    let js_code = format!(
        "navigator.clipboard.writeText('{}').catch(e => console.warn('Copy failed:', e))",
        text.replace('\'', "\\'")
    );
    let _ = js_sys::eval(&js_code);
}

File: crates/ifc-lite-swift/IfcLiteViewer/IfcLiteViewer/Components/PropertiesPanel.swift
import SwiftUI
import IfcLite

struct PropertiesPanel: View {
    @EnvironmentObject var viewModel: ViewerViewModel

    var body: some View {
        VStack(spacing: 0) {
            // Header
            HStack {
                Text("Properties")
                    .font(.headline)
                Spacer()
            }
            .padding(.horizontal)
            .padding(.vertical, 8)

            Divider()

            if let entity = viewModel.selectedEntity {
                ScrollView {
                    VStack(alignment: .leading, spacing: 16) {
                        // Entity Info Section
                        PropertySection(title: "Entity Info") {
                            PropertyRow(label: "Type", value: entity.entityType)
                            if let name = entity.name {
                                PropertyRow(label: "Name", value: name)
                            }
                            PropertyRow(label: "ID", value: "#\(entity.id)")
                            if let globalId = entity.globalId {
                                PropertyRow(label: "GlobalId", value: globalId, copyable: true)
                            }
                            if let storey = entity.storey {
                                PropertyRow(label: "Storey", value: storey)
                            }
                            if let elevation = entity.storeyElevation {
                                PropertyRow(label: "Elevation", value: String(format: "%.2f m", elevation))
                            }
                        }

                        // Actions Section
                        PropertySection(title: "Actions") {
                            ActionButtonsView(entityId: entity.id)
                        }

                        // Properties Section
                        let propertySets = viewModel.scene.getProperties(entityId: entity.id)
                        ForEach(Array(propertySets.enumerated()), id: \.offset) { index, pset in
                            PropertySection(title: pset.name) {
                                ForEach(Array(pset.properties.enumerated()), id: \.offset) { pIndex, prop in
                                    PropertyRow(
                                        label: prop.name,
                                        value: prop.unit != nil ? "\(prop.value) \(prop.unit!)" : prop.value
                                    )
                                }
                            }
                        }
                    }
                    .padding()
                }
            } else if viewModel.selectedIds.count > 1 {
                MultiSelectionView()
            } else {
                NoSelectionView()
            }
        }
        .background(Color(white: 0.15))
    }
}

struct PropertySection<Content: View>: View {
    let title: String
    @ViewBuilder let content: Content

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(title)
                .font(.subheadline)
                .fontWeight(.semibold)
                .foregroundColor(.secondary)

            VStack(alignment: .leading, spacing: 4) {
                content
            }
            .padding(12)
            .background(Color.secondary.opacity(0.1))
            .cornerRadius(8)
        }
    }
}

struct PropertyRow: View {
    let label: String
    let value: String
    var copyable: Bool = false

    var body: some View {
        HStack {
            Text(label)
                .foregroundColor(.secondary)
            Spacer()
            Text(value)
                .lineLimit(1)
            if copyable {
                Button(action: copyToClipboard) {
                    Image(systemName: "doc.on.doc")
                        .font(.caption)
                }
                .buttonStyle(.borderless)
            }
        }
        .font(.caption)
    }

    private func copyToClipboard() {
        #if os(macOS)
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(value, forType: .string)
        #else
        UIPasteboard.general.string = value
        #endif
    }
}

struct ActionButtonsView: View {
    @EnvironmentObject var viewModel: ViewerViewModel
    let entityId: UInt64

    var body: some View {
        VStack(spacing: 8) {
            HStack(spacing: 8) {
                ActionButton(title: "Zoom to", icon: "magnifyingglass") {
                    viewModel.zoomToEntity(entityId)
                }

                ActionButton(title: "Isolate", icon: "eye.circle") {
                    viewModel.isolateEntity(entityId)
                }
            }

            HStack(spacing: 8) {
                ActionButton(title: "Hide", icon: "eye.slash") {
                    viewModel.hideEntity(entityId)
                }

                ActionButton(title: "Select Similar", icon: "doc.on.doc") {
                    selectSimilar()
                }
            }
        }
    }

    private func selectSimilar() {
        guard let entity = viewModel.entities.first(where: { $0.id == entityId }) else { return }
        let similarIds = viewModel.entities
            .filter { $0.entityType == entity.entityType }
            .map { $0.id }
        for id in similarIds {
            viewModel.selectedIds.insert(id)
        }
    }
}

struct ActionButton: View {
    let title: String
    let icon: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack {
                Image(systemName: icon)
                Text(title)
                    .font(.caption)
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, 8)
            .background(Color.secondary.opacity(0.15))
            .cornerRadius(6)
        }
        .buttonStyle(.plain)
    }
}

struct NoSelectionView: View {
    var body: some View {
        VStack(spacing: 16) {
            Image(systemName: "cursorarrow.click")
                .font(.system(size: 36))
                .foregroundColor(.secondary)
            Text("No Selection")
                .font(.headline)
            Text("Click on an element in the 3D view or hierarchy to see its properties.")
                .font(.subheadline)
                .foregroundColor(.secondary)
                .multilineTextAlignment(.center)
        }
        .padding()
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

struct MultiSelectionView: View {
    @EnvironmentObject var viewModel: ViewerViewModel

    var body: some View {
        VStack(spacing: 16) {
            Image(systemName: "checkmark.circle.fill")
                .font(.system(size: 36))
                .foregroundColor(.accentColor)

            Text("\(viewModel.selectedIds.count) items selected")
                .font(.headline)

            VStack(spacing: 8) {
                Button("Isolate All") {
                    viewModel.isolateSelection()
                }
                .buttonStyle(.borderedProminent)

                Button("Hide All") {
                    for id in viewModel.selectedIds {
                        viewModel.hideEntity(id)
                    }
                }
                .buttonStyle(.bordered)

                Button("Clear Selection") {
                    viewModel.clearSelection()
                }
                .buttonStyle(.bordered)
            }
        }
        .padding()
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

#Preview {
    PropertiesPanel()
        .environmentObject(ViewerViewModel())
        .frame(width: 280)
}
